#13 - Book and Category Testing Implementation

waspp01 opened the pull request:
- Created tests for Service and Controller layers for Book and Category entities
 - Other minor fixes Coverage: https://drive.google.com/file/d/1qqvA7cOKn6UHY3HVSPqI1qcxDSVRlky_/view?usp=sharing

waspp01 committed:
#13 - Book and Category Testing Implementation
 - Created tests for Service and Controller layers for Book and Category entities
 - Other minor fixes
 Coverage:
 https://drive.google.com/file/d/1qqvA7cOKn6UHY3HVSPqI1qcxDSVRlky_/view?usp=sharing

Sanderfuck commented on src/test/java/com/example/onlinebookstore/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,294 @@
+            connection.setAutoCommit(true);
+            ScriptUtils.executeSqlScript(
+                    connection,
+                    new ClassPathResource("database/books/insert-test-books.sql")
+            );
```
```suggestion
                    new ClassPathResource("database/books/insert-test-books.sql"));
```

Sanderfuck commented on src/test/java/com/example/onlinebookstore/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,294 @@
+            connection.setAutoCommit(true);
+            ScriptUtils.executeSqlScript(
+                    connection,
+                    new ClassPathResource("database/books/delete-all-test-books.sql")
+            );
+
```
```suggestion
                    new ClassPathResource("database/books/delete-all-test-books.sql"));
```

Sanderfuck commented on src/test/java/com/example/onlinebookstore/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,294 @@
+        //then
+        BookDto actual = objectMapper.readValue(
+                result.getResponse().getContentAsString(), BookDto.class
+        );
```
Don`t replace it for next line, it looks not good

Sanderfuck commented on src/test/java/com/example/onlinebookstore/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,294 @@
+            """)
+    void createBook_ValidCreateBookRequestDto_ReturnsBookDto() throws Exception {
+        //given
+        CreateBookRequestDto createBookRequestDto = new CreateBookRequestDto()
+                .setTitle("TestTitle")
+                .setAuthor("TestAuthor")
+                .setIsbn("TestIsbn")
+                .setCategoryIds(Set.of(1L))
+                .setPrice(BigDecimal.TEN)
+                .setDescription("Test Description")
+                .setCoverImage("test cover");
+
+        BookDto expected = new BookDto()
+                .setIsbn(createBookRequestDto.getIsbn())
+                .setDescription(createBookRequestDto.getDescription())
+                .setAuthor(createBookRequestDto.getAuthor())
+                .setPrice(createBookRequestDto.getPrice())
+                .setCoverImage(createBookRequestDto.getCoverImage())
+                .setCategoryIds(createBookRequestDto.getCategoryIds())
+                .setTitle(createBookRequestDto.getTitle());
```
Create templates and reuse it in tests

Sanderfuck commented on src/test/java/com/example/onlinebookstore/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,279 @@
+    @Test
+    @WithMockUser
+    @Sql(
+            scripts = {
+                    "classpath:database/books-categories/insert-books-categories.sql",
+                    "classpath:database/books/insert-test-books.sql"
+            },
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD
+    )
```
Place brackets in the same line where code ends

Sanderfuck replied:
fix it everywhere

Sanderfuck commented on src/test/java/com/example/onlinebookstore/service/BookServiceImplTest.java:
```diff
@@ -0,0 +1,315 @@
+
+@ExtendWith(MockitoExtension.class)
+public class BookServiceImplTest {
+    @InjectMocks
+    private BookServiceImpl bookServiceImpl;
```
InjectMocks field place in the end of fields

Sanderfuck commented on src/test/java/com/example/onlinebookstore/service/BookServiceImplTest.java:
```diff
@@ -0,0 +1,315 @@
+        BookSearchParameters bookSearchParameters =
+                new BookSearchParameters(new String[0], new String[0]);
+
+        Pageable pageable = PageRequest.of(0,10);
```
```suggestion
        Pageable pageable = PageRequest.of(0, 10);
```

Sanderfuck replied:
Use Ctrl+Alt+L in this case

Sanderfuck commented on src/test/java/com/example/onlinebookstore/service/BookServiceImplTest.java:
```diff
@@ -0,0 +1,315 @@
+        bookDto.setDescription(book.getDescription());
+        bookDto.setCoverImage(book.getCoverImage());
+        bookDto.setCategoryIds(book.getCategories().stream()
+                .map(c -> c.getId())
+                .collect(Collectors.toSet()));
+        return bookDto;
+    }
+
+    private Book getTestBook() {
+        Book book = new Book();
+        book.setId(1L);
+        book.setTitle("TestBook1");
+        book.setAuthor("TestAuthor1");
+        book.setIsbn("TestIsbn1");
+        book.setPrice(BigDecimal.valueOf(9.99));
+        book.setDescription("Test Description1");
+        book.setCoverImage("TestCoverImage");
+        book.setCategories(Set.of(getTestCategory()));
+        return book;
+    }
+
+    private Category getTestCategory() {
+        Category category = new Category();
+        category.setId(1L);
+        category.setName("Test Name");
+        category.setDescription("Test Description");
+        return category;
+    }
+
+    private BookDtoWithoutCategoryIds getBookDtoWithoutCategoryIdsFromBook(Book book) {
+        BookDtoWithoutCategoryIds dto = new BookDtoWithoutCategoryIds();
+        dto.setId(book.getId());
+        dto.setAuthor(book.getAuthor());
+        dto.setIsbn(book.getIsbn());
+        dto.setTitle(book.getTitle());
+        dto.setDescription(book.getDescription());
+        dto.setCoverImage(book.getCoverImage());
+        dto.setPrice(book.getPrice());
+        return dto;
+    }
```
If you use this templates in other tests classes, you can make it static and place in some util class

waspp01 committed:
#13 - Book and Category Testing Implementation
 - Fixed checkstyle issues
 - Added TestBookSupplier and TestCategorySupplier classes for unification
 - Improved BookControllerTest class tests consistency
